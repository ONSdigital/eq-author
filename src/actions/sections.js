export const ADD_SECTION = 'ADD_SECTION';
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER = 'ADD_ANSWER';

export const addSection = name => {
  return {
    type: ADD_SECTION,
    payload: {
      name: name,
    },
  };
};

export const addQuestion = (name, sectionID) => {
  return {
    type: ADD_QUESTION,
    payload: {
      name: name,
      sectionID: sectionID,
    },
  };
};

export const addAnswer = (name, questionID, sectionID) => {
  return {
    type: ADD_ANSWER,
    payload: {
      name: name,
      questionID: questionID,
      sectionID: sectionID,
    },
  };
};
