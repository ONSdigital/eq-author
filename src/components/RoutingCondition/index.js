import React from "react";
import Select from "components/Forms/Select";
import styled from "styled-components";

import IconButton from "components/IconButton";
import IconClose from "./icon-close.svg?inline";
import { PropTypes } from "prop-types";
import CustomPropTypes from "custom-prop-types";

import { Grid, Column } from "components/Grid";

import svgPath from "./path.svg";
import svgPathEnd from "./path-end.svg";

const Label = styled.label`
  width: 100%;
  display: inline-block;
  font-size: 0.9em;
  font-weight: 600;
  text-align: center;
  align-self: center;
`;

export const PageSelect = styled(Select)`
  margin: 0;
  align-self: center;
`;

export const RemoveButton = styled(IconButton)`
  margin: auto;
  align-self: center;
`;

const ConnectedPath = styled.div`
  position: relative;
  height: 100%;

  &::after {
    position: absolute;
    content: "";
    background: url(${({ pathEnd }) => (pathEnd ? svgPathEnd : svgPath)})
      no-repeat center center;
    background-size: auto;
    width: 100%;
    height: calc(100% - 2em);
    top: 0;
    bottom: 0;
    margin: auto;
  }
`;

const RoutingCondition = ({
  sections,
  value,
  selectedPage,
  id,
  pathEnd,
  onPageChange,
  onRemoveClick,
  children
}) => (
  <div>
    <Grid align="center">
      <Column gutters={false} cols={1.5}>
        <Label htmlFor={id}>IF</Label>
      </Column>
      <Column gutters={false} cols={9}>
        <PageSelect
          defaultValue={selectedPage.id}
          value={value}
          onChange={onPageChange}
          id={id}
        >
          {sections.map(section => (
            <optgroup label={section.title} key={section.id}>
              {section.pages.map(page => (
                <option value={page.id} key={page.id} disabled={page.disabled}>
                  {page.title}
                </option>
              ))}
            </optgroup>
          ))}
        </PageSelect>
      </Column>
      <Column gutters={false} cols={1.5}>
        <RemoveButton icon={IconClose} onClick={onRemoveClick} iconOnly>
          Remove
        </RemoveButton>
      </Column>
    </Grid>
    <Grid>
      <Column gutters={false} cols={1.5}>
        <ConnectedPath pathEnd={pathEnd} />
      </Column>
      <Column gutters={false} cols={9}>
        {children}
      </Column>
      <Column cols={1.5} />
    </Grid>
  </div>
);

RoutingCondition.propTypes = {
  sections: PropTypes.arrayOf(CustomPropTypes.section).isRequired,
  value: PropTypes.string,
  selectedPage: CustomPropTypes.page.isRequired,
  id: PropTypes.string.isRequired,
  onPageChange: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  pathEnd: PropTypes.bool.isRequired
};

RoutingCondition.defaultProps = {
  pathEnd: false
};

export default RoutingCondition;
