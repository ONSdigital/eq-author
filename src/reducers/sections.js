import _, {merge} from 'lodash';
import * as actions from 'actions/sections';

const ids = [];
const getIdFromName = name => {
  const newId = name.replace(/\s+/g, '-').toLowerCase();
  ids.push(newId);
  return newId;
};

const getSectionById = (sections, id) =>
  _(sections).filter({id: id}).value()[0];

const sections = (state = {sections: []}, action) => {
  switch (action.type) {
    case actions.ADD_SECTION:
      return {
        sections: [
          ...state.sections,
          {
            id: getIdFromName(action.payload.name),
            name: action.payload.name,
            questions: [],
          },
        ],
      };

    case actions.ADD_QUESTION:
      return merge(
        {},
        state.sections,
        getSectionById(
          state.sections,
          action.payload.sectionID
        ).questions.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
          answers: [],
        })
      );

    case actions.ADD_ANSWER:
      _(state.sections)
        .map('questions')
        .flatten()
        .filter({id: action.payload.questionID})
        .value()[0]
        .answers.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
        });

      return merge(
        {},
        state.sections,
        getSectionById(state.sections, action.payload.sectionID)
      );

    default:
      return state;
  }
};

export default sections;
