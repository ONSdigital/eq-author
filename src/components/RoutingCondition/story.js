/* eslint-disable react/prop-types, jsx-a11y/href-no-hash */

import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import styled from "styled-components";
import { first } from "lodash";

import RoutingCondition from "components/RoutingCondition";
import MultipleChoiceAnswerOptionsSelector from "./MultipleChoiceAnswerOptionsSelector";
import { Alert, AlertText, AlertTitle } from "./Alert";

import sections from "./mockstate";

const Background = styled.span`
  padding: 1em;
  display: block;
  max-width: 55em;
`;

const selectedPage = first(first(sections).pages);

storiesOf("RoutingCondition", module)
  .addDecorator(story => <Background>{story()}</Background>)
  .add("Default", () => (
    <RoutingCondition
      id="routing-condition"
      selectedPage={selectedPage}
      onPageChange={action("Page selected")}
      onRemoveClick={action("Close")}
      sections={sections}
    >
      <MultipleChoiceAnswerOptionsSelector
        options={selectedPage.options}
        onOptionSelectionChange={action("Option selected")}
        onSelectAll={action("Select all")}
      />
    </RoutingCondition>
  ))
  .add("No answers", () => (
    <RoutingCondition
      id="routing-condition"
      selectedPage={selectedPage}
      onPageChange={action("Page selected")}
      onRemoveClick={action("Close")}
      sections={sections}
      pathEnd
    >
      <Alert>
        <AlertTitle>
          No answers have been added to this question yet.
        </AlertTitle>
        <AlertText>
          First, <a href="#">add an answer</a> to continue.
        </AlertText>
      </Alert>
    </RoutingCondition>
  ))
  .add("Routing unsupported", () => (
    <RoutingCondition
      id="routing-condition"
      selectedPage={selectedPage}
      onPageChange={action("Page selected")}
      onRemoveClick={action("Close")}
      sections={sections}
      pathEnd
    >
      <Alert>
        <AlertTitle>
          Routing is not available for this type of answer
        </AlertTitle>
        <AlertText>
          You cannot route on &apos;date range&apos; answers
        </AlertText>
      </Alert>
    </RoutingCondition>
  ));
