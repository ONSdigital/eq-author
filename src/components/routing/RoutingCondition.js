import React from "react";
import styled from "styled-components";

import DeleteButton from "components/DeleteButton";
import IconClose from "./icon-close.svg?inline";
import { PropTypes } from "prop-types";
import CustomPropTypes from "custom-prop-types";

import { Grid, Column } from "components/Grid";
import { NavLink } from "react-router-dom";

import svgPath from "./path.svg";
import svgPathEnd from "./path-end.svg";
import IconText from "components/IconText";
import {
  get,
  isNil,
  isEmpty,
  lowerCase,
  uniqueId,
  first,
  flow,
  negate,
  overSome
} from "lodash";
import MultipleChoiceAnswerOptionsSelector from "components/routing/MultipleChoiceAnswerOptionsSelector";
import GroupedSelect from "./GroupedSelect";
import Transition from "components/routing/Transition";
import { TransitionGroup } from "react-transition-group";
import { Alert, AlertTitle, AlertText } from "components/routing/Alert";
import { buildPagePath } from "utils/UrlUtils";
import isAnswerValidForRouting from "./isAnswerValidForRouting";

import routingConditionFragment from "graphql/fragments/routing-condition.graphql";

const Label = styled.label`
  width: 100%;
  display: inline-block;
  font-size: 0.9em;
  letter-spacing: 0.05em;
  font-weight: bold;
  text-align: center;
  align-self: center;
`;

export const PageSelect = styled(GroupedSelect).attrs({
  onChange: props => ({ value }) => props.onChange(value)
})`
  margin: 0;
  align-self: center;
`;

const ConnectedPath = styled.div`
  position: relative;
  height: 100%;

  &::after {
    position: absolute;
    content: "";
    background: url(${({ pathEnd }) => (pathEnd ? svgPathEnd : svgPath)})
      no-repeat center center;
    background-size: auto;
    width: 100%;
    height: calc(100% - 2em);
    top: 0;
    bottom: 0;
    margin: auto;
  }
`;

const RemoveButton = styled(DeleteButton)`
  display: block;
  margin: auto;
  position: relative;
  right: 2px;
`;

const firstAnswerIsValid = flow(
  first,
  isAnswerValidForRouting
);
const shouldDisable = overSome([isEmpty, negate(firstAnswerIsValid)]);

const convertToGroups = sections =>
  sections.map(section => ({
    label: section.displayName,
    id: section.id,
    options: section.pages.map(page => ({
      label: page.displayName,
      value: page.id,
      disabled: shouldDisable(page.answers)
    }))
  }));

const renderNoAnswer = params => (
  <Transition exit={false}>
    <Alert data-test="no-answer-msg">
      <AlertTitle>No answers have been added to this question yet.</AlertTitle>
      <AlertText>
        First, <NavLink to={buildPagePath(params)}>add an answer</NavLink> to
        continue.
      </AlertText>
    </Alert>
  </Transition>
);

const renderUnsupportedAnswer = answer => (
  <Transition key="answer" exit={false}>
    <Alert data-test="invalid-answer-type-msg">
      <AlertTitle>Routing is not available for this type of answer</AlertTitle>
      <AlertText>
        You cannot route on &lsquo;
        {lowerCase(answer.type)}
        &rsquo; answers.
      </AlertText>
    </Alert>
  </Transition>
);

const renderDeletedQuestion = () => (
  <Transition key="answer" exit={false}>
    <Alert data-test="deleted-answer-msg">
      <AlertTitle>
        The question this condition referred to has been deleted
      </AlertTitle>
      <AlertText>
        Please select a new question from the dropdown above.
      </AlertText>
    </Alert>
  </Transition>
);

const renderCannotAddAndCondition = () => (
  <Transition key="answer" exit={false}>
    <Alert data-test="and-not-valid-msg">
      <AlertTitle>
        AND condition not valid with &lsquo;radio button&rsquo; answer
      </AlertTitle>
      <AlertText>Please select a different question.</AlertText>
    </Alert>
  </Transition>
);

const renderEditor = (condition, onToggleOption) => (
  <Transition key="answer" exit={false}>
    <MultipleChoiceAnswerOptionsSelector
      condition={condition}
      onOptionSelectionChange={onToggleOption}
    />
  </Transition>
);

const RoutingCondition = ({
  condition,
  ruleId,
  sections,
  label,
  onPageChange,
  onRemove,
  onToggleOption,
  canAddAndCondition,
  match
}) => {
  let editor;
  let value = get(condition, "questionPage.id");
  let pageSelectIsValid;

  if (isNil(condition.questionPage)) {
    editor = renderDeletedQuestion();
  } else if (isNil(condition.answer)) {
    value = null;
    pageSelectIsValid = false;
    editor = renderNoAnswer(match.params);
  } else if (!isAnswerValidForRouting(condition.answer)) {
    value = null;
    pageSelectIsValid = false;
    editor = renderUnsupportedAnswer(condition.answer);
  } else if (!canAddAndCondition) {
    pageSelectIsValid = false;
    editor = renderCannotAddAndCondition();
  } else {
    editor = renderEditor(condition, onToggleOption);
  }

  const id = uniqueId("RoutingCondition");
  const handleRemove = onRemove ? () => onRemove(ruleId, condition.id) : null;
  const handleChange = ({ value }) =>
    onPageChange({ id: condition.id, questionPageId: value });

  return (
    <div data-test="routing-condition">
      <Grid align="center">
        <Column gutters={false} cols={1}>
          <Label htmlFor={id}>{label}</Label>
        </Column>
        <Column gutters={false} cols={10}>
          <PageSelect
            value={value}
            valid={pageSelectIsValid}
            onChange={handleChange}
            groups={convertToGroups(sections)}
            id={id}
          />
        </Column>
        <Column gutters={false} cols={1}>
          <RemoveButton
            onClick={handleRemove}
            disabled={!onRemove}
            data-test="btn-remove"
          >
            <IconText icon={IconClose} hideText>
              Remove
            </IconText>
          </RemoveButton>
        </Column>
      </Grid>
      <Grid>
        <Column gutters={false} cols={1}>
          <ConnectedPath pathEnd={isNil(condition.answer)} />
        </Column>
        <Column gutters={false} cols={10}>
          <TransitionGroup>{editor}</TransitionGroup>
        </Column>
        <Column cols={1} />
      </Grid>
    </div>
  );
};

RoutingCondition.propTypes = {
  ruleId: PropTypes.string.isRequired,
  condition: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  sections: PropTypes.arrayOf(CustomPropTypes.section).isRequired,
  onPageChange: PropTypes.func.isRequired,
  onToggleOption: PropTypes.func.isRequired,
  onRemove: PropTypes.func,
  label: PropTypes.oneOf(["IF", "AND"]).isRequired,
  match: CustomPropTypes.match,
  canAddAndCondition: PropTypes.bool.isRequired
};

RoutingCondition.defaultProps = {
  label: "IF"
};

RoutingCondition.fragments = {
  RoutingCondition: routingConditionFragment
};

export default RoutingCondition;
