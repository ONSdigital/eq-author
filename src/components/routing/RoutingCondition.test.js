import React from "react";
import { shallow } from "enzyme";

import RoutingCondition, { RemoveButton, PageSelect } from "./RoutingCondition";

import sections from "./mockstate";

let wrapper, mockHandlers;

describe("components/RoutingCondition", () => {
  beforeEach(() => {
    mockHandlers = {
      onPageChange: jest.fn(),
      onRemoveClick: jest.fn()
    };

    wrapper = shallow(
      <RoutingCondition
        id="test"
        sections={sections}
        selectedPage={sections[0].pages[0]}
        {...mockHandlers}
      >
        <div>I am a child</div>
      </RoutingCondition>
    );
  });

  it("should render consistently", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onRemoveClick handler when remove button is clicked", () => {
    wrapper.find(RemoveButton).simulate("click");
    expect(mockHandlers.onRemoveClick).toHaveBeenCalled();
  });

  it("should call onUpdate with value of selected page", () => {
    const arg = { value: "hello" };
    wrapper.find(PageSelect).simulate("change", arg);
    expect(mockHandlers.onPageChange).toHaveBeenCalledWith(arg);
  });
});
