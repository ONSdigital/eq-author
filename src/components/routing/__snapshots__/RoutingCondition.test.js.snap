// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RoutingCondition should render consistently 1`] = `
<div>
  <Grid
    align="center"
    fillHeight={true}
  >
    <Column
      cols={1}
      gutters={false}
    >
      <RoutingCondition__Label
        htmlFor="test"
      >
        IF
      </RoutingCondition__Label>
    </Column>
    <Column
      cols={10}
      gutters={false}
    >
      <RoutingCondition__PageSelect
        defaultValue="2"
        id="test"
        onChange={[Function]}
      >
        <optgroup
          key="1"
          label="1. Integer posuere erat a ante"
        >
          <option
            disabled={false}
            key="2"
            value="2"
          >
            1.1 Curabitur blandit tempus porttitor
          </option>
          <option
            disabled={false}
            key="9"
            value="9"
          >
            1.2 Nulla vitae elit libero, a pharetra augue.
          </option>
        </optgroup>
      </RoutingCondition__PageSelect>
    </Column>
    <Column
      cols={1}
      gutters={false}
    >
      <RoutingCondition__RemoveButton
        disabled={false}
        onClick={[Function]}
      >
        <IconText
          hideText={true}
          icon={[Function]}
        >
          Remove
        </IconText>
      </RoutingCondition__RemoveButton>
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={1}
      gutters={false}
    >
      <RoutingCondition__ConnectedPath
        pathEnd={false}
      />
    </Column>
    <Column
      cols={10}
      gutters={false}
    >
      <div>
        I am a child
      </div>
    </Column>
    <Column
      cols={1}
    />
  </Grid>
</div>
`;
